rfcs/text/0000-planning-for-the-dcp.md

```diff
@@ -0,0 +1,211 @@
+After this meeting, the **Product Owner** and **Technical Lead** for each component MUST ensure that the *Product Backlog* in their repo is refined and aligned to the DCP priorities.
+
+(See *Refinement* in [*Scrum Guide*](https://scrumguides.org/docs/scrumguide/v2017/2017-Scrum-Guide-US.pdf) and [*Nexus Guide*](http://scrumorg-website-prod.s3.amazonaws.com/drupal/2018-01/2018-Nexus-Guide-English_0.pdf) for guidance.)
+
+#### Refinement and ZenHub Pipelines
+
+New items in the *dcp* repo appear in the ZenHub *New* pipeline and MUST be triaged into the *Product Backlog* pipeline, the *Icebox* pipeline, or the *Closed* pipeline.
+
+The ZenHub *Product Backlog* pipeline is the top-level DCP product backlog which is ordered during Refinement.
+
+The ZenHub *Icebox* pipeline SHOULD be used with great constraint in the *dcp* repo. It's not a pipeline for avoiding hard decisions during Refinement. It's intended to document potential *futures* from the DCP roadmap. Individual component repo(s) MAY use the *Icebox* pipeline in a manner of their own choosing.
+
+No issues in the *dcp* repo MUST be assigned to the *Epic* pipeline. Individual component repo(s) MAY use the *Epic* pipeline in a manner of their own choosing.
+
+### Sprint Planning
+
+Prior to the start of the DCP Sprint, component teams MUST identify their *Sprint Backlog* - a slice of items from the the ZenHub *Product Backlog* pipeline in their repo which are assigned to the ZenHub Milestone for the Sprint and moved to the ZenHub *Sprint Backlog* pipeline.
+
+The number of items selected is solely up to the Development Team. 
+- [ZenHub story points](https://help.zenhub.com/support/solutions/articles/43000034282) MUST be assigned to all *Sprint Backlog* items. (Story points are required to enable [ZenHub burndown charts](https://help.zenhub.com/support/solutions/articles/43000010356-track-sprint-progress-with-burndown-charts) to track the progress of a Sprint.)
+  - [How to start?](https://www.scrum.org/forum/scrum-forum/6774/how-decide-how-many-story-points-take-first-sprint)
+- Research and prototyping tasks SHOULD be modeled as Spikes.
+- Dependencies between components within a Sprint MUST be communicated between the respective teams and surfaced using [ZenHub Dependencies](https://help.zenhub.com/support/solutions/articles/43000010349). 
+
+At the start of the Sprint, the single DCP Sprint Backlog SHOULD be reviewed by the **PM Team** and **Architecture Team** in their oversight roles.
+
+### Sprint(ing)
+
+During Development, individual developers assign items from the ZenHub *Sprint Backlog* pipeline and move to the *In Progress* pipeline. 
+
+When there is a pull request in review for an *In Progress* item, the item is moved to the *QA/Review* pipeline. 
+
+If rejected, the item returns to *In Progress*; otherwise, it is deployed in Dev for verification and moves to the *Done* pipeline.
+
+Items that are *Done* are moved to the *Closed* pipeline when verified in Dev or Integration. 
```

Comment:
I propose an extension to that. Consider adding the following pipelines to Zenhub:

* [Preceding pipelines as currently specified in this RFC]
* In Progress — PR in the works
* Review — PR filed and review requested
* Merged — PR approved and merged to default branch (develop or master for most repos) and deployed to component's `dev` deployment
* Integrated — Changes promoted to `integration` deployment
* Staged — Changes promoted to `staging` deployment
* Deployed — Changes promoted to `prod` deployment
* Closed — Changes (feature or bug) verified by original author in production

This makes it very easy to tell at a glance how far a particular change has progressed through the promotion sequence. It makes it easy for ticket stake holders to examine the effect of a change in deployments preceding production.

Reply:
If there is consensus among your peer Technical Leads to extend the pipelines for better transparency on the sequence, then I'm happy to make the change. It does require more developer discipline to succeed.

Reply:
We could specify that teams SHOULD makes use of it but MAY skip pipelines and go directly from Review to Closed. There are ticket types that don't manifest in code changes. We label them as `task` in our repo. These task tickets should go directly from in Progress to Closed.

Reply:
What are the next steps to get this into the draft? I don't think we should be doing a mini-RFC process on this proposal. If the author doesn't disagree, they should work it in, resolve the comment and the usual "silence is consent" approach should apply.

Reply:
@hannes-ucsc - Not based on my experience as a technical editor and author at the IETF. You wait during the time-boxed review period to see if there's rough consensus supporting a change. I'd like to hear from other leads on this one before making the change.

Reply:
Also - I prefer to avoid *SHOULD* because it's often an easy way to add complexity to a design for one requestor and avoid a consensus conversation because everyone who disagrees knows that they can safely ignore it.

Reply:
Where exactly do you see the added complexity in allowing ticket to bypass the new  pipelines I proposed? In fact, I think making them mandatory adds complexity for things like tasks where the added pipelines don't make sense.

Reply:
Regarding process: Assuming we resolve the SHOULD/MUST issue, is it your intention to merge this unless you hear dissent from other leads?

Reply:
I think having integrated, staged and deployed pipelines could be good. I'm often confused as to the progress through environments of a particular feature. But @hannes-ucsc, don't 'In Progress', 'Review' and 'Merged' already all exist just with slightly different names?

Reply:
Correct, @justincc. I edited my comment.

Reply:
@hannes-ucsc - See my proposed changes below. I suggest that we use the names of the stages from the SOP ("Production" rather than "Deployed") to match the documentation which I've now referenced in the RFC.

During Development, the Development Team assigns issues from the ZenHub *Sprint Backlog* pipeline and move to the *In Progress* pipeline. 

When there is a pull request filed and in review for an *In Progress* item, the issue is moved to the *Review* pipeline. 

If the pull request is rejected, the issue returns to *In Progress*.

Otherwise, the pipeline tracks the issue through the DCP [promotion sequence](https://allspark.dev.data.humancellatlas.org/dcp-ops/docs/wikis/SOP:%20Releasing%20new%20Versions%20of%20DCP%20Software).

The approved pull request is merged to the default branch, deployed to Development, and the issue moves to the *Development* pipeline.

When the changes are promoted to Integration, the issue moves to the *Integration* pipeline.

When the changes are promoted to Staging, the issue moves to the *Staging* pipeline.

When the changes are promoted to Production, the issue moves to the *Production* pipeline.

When the original author verifies the changes in Production, the issue then moves to the *Closed* pipeline.

Reply:
I'm fine with the name changes. The Development pipeline may be confused with the In Progress pipeline. In my team, there is a little bit of a delay between when a PR is merged in to the default branch and the time the default branch is deployed to our `dev` environment. But we can deal with that wrinkle with more automation or labels.

Reply:
There are other reasons to have tasks in review. For example, when we complete an ingest feature we need wranglers to review it before we can consider it done.